bokchoi/gcp/gcp.py

```diff
@@ -0,0 +1,262 @@
+                    'boot': True,
+                    'autoDelete': True,
+                    'initializeParams': {
+                        'sourceImage': image_response['selfLink'],
+                        'diskSizeGb': self.gcp.get('disk_space')
+                    }
+                }
+            ],
+            'scheduling': {
+              'preemptible': self.gcp.get('preemptible')
+            },
+
+            'networkInterfaces': [{
+                'network': 'global/networks/{}'.format(self.gcp.get('network')),
+                'subnetwork': '/regions/{}/subnetworks/{}'.format(
+                    self.gcp.get('region'), self.gcp.get('sub_network')),
+                'accessConfigs': [
+                    {'type': 'ONE_TO_ONE_NAT', 'name': 'External NAT'}
+                ]
+            }],
+
+            'serviceAccounts': [{
+                'email': 'default',
+                'scopes': [
+                    'https://www.googleapis.com/auth/devstorage.read_write',
+                    'https://www.googleapis.com/auth/logging.write'
+                ]
+            }],
+
+            'metadata': {
+                'items': [{
+                    'key': 'startup-script',
+                    'value': startup_script
+                }, {
+                    'key': 'bucket_name',
+                    'value': self.gcp.get('bucket')
+                }, {
+                    'key': 'package_name',
+                    'value': '{}-{}.zip'.format(self.project_name, 'package')
+                }, {
+                    'key': 'entry_point',
+                    'value': self.entry_point
+                }]
+            }
+        }
+
+        return config
+
+    def create_instance(self):
+        """Create a new compute engine"""
+        print('Creating instance')
+        try:
+            return self.compute.instances().insert(
+                project=self.gcp.get('project'),
+                zone=self.gcp.get('zone'),
+                body=self.define_instance_config()).execute()
+        except googleapiclient.errors.HttpError as e:
+            if 'already exists' in str(e):
+                print('instance with name {} already exists. exit(1)'.format(self.project_name))
+                sys.exit(1)
+            else:
+                print(e)
+                sys.exit(1)
+
+    def delete_instance(self):
+        """Remove the created compute engine"""
+        print('Deleting instance')
+        return self.compute.instances().delete(
+            project=self.gcp.get('project'),
+            zone=self.gcp.get('zone'),
+            instance=self.project_name).execute()
+
+    def wait_for_operation(self, operation):
+        """Method which polls the status of the operations and returns when the
+        operation is completed.
+        :arg operation: a gcp api operation
+        """
+        if operation is None:
+            return
+
+        print('Waiting for operation to finish...')
+        while True:
```

Comment:
There is a retry function in bokchoi.utils that might be useful

Reply:
all GCP operations are async and this function blocks the current process until the operation has been succeeded (e.g. make the create instance operation blocking). I wouldn't say that it's the same thing as retrying an operation due to Errors (as described in the utils.retry function).